jQuery Validation plugin implementation

stevehobbsdev opened the pull request:
This is a complete implementation of the jQuery Validation plugin, as seen at http://docs.jquery.com/Plugins/Validation.

One thing I have experimented with this particular plugin is how it is called. When developing import libraries internally which map to jQuery plugins, we have preferred to call the plugin passing in the element which should be acted upon, rather than inheriting from jQueryObject and attempting to use jQuery in ScriptSharp as we would in plain Javascript. This removed the need to cast jQueryObjects when calling plugins from ScriptSharp.

In this case, instead of the call being

`````` ((jQueryValidation)jQuery.Select("form")).Validate()```

we can instead call:
```jQueryValidation.Validate(jQuery.Select("form"))```

This is done by asking ScriptSharp to compile the method as `$.fn.validate.call` using a `ScriptAliasAttribute`, which allows us to call a jQuery plugin method by passing the selector in.

Please do let me know your thoughts on this approach.
``````

Commit:
Moved toJS() and toJSON() from KnockoutMapping to Knockout to reflect actual API

Commit:
Merge branch 'Validator' of https://github.com/elkdanger/scriptsharp into Validator

Commit:
Added support for rules and validation methods

Commit:
Added type to represent validation methods

Commit:
Reverted jQuery change from branch

Commit:
Revert "Moved toJS() and toJSON() from KnockoutMapping to Knockout to reflect actual API"

This reverts commit 15556a72b8e4ad36e52bde4c5bce05813b325bb5.

Commit:
Added classes for managing validation methods and messages

Commit:
Added xml documentation for jQueryValidation

Commit:
Added types for messages and validation rules

nikhilk commented:
Actually the model isn't to cast, but rather use the following pattern:

```
jQuery.Select("form").Plugin<ValidationObject>().Validate();
```

where ValidationObject is a class deriving from jQueryObject (so it supports chaining like any other jQuery API, if Validate on the plugin follows the general model as any other jQueryObject).

Did you not like that pattern?

To be clear, I don't love that pattern either. It is there to make up for the lack of extension method support, since extension methods are c# 3.0 and script# restricts to c# 2.0 syntax at the moment. I'd expect in the future that plugin methods are imported in the form of extension methods and the usage model would be the following:

```
jQuery.Select("form").Validate();
```

Matter of opinion, but I thought the generic trick using the `Plugin<TPlugin>()` method was closer to the intended final state.

A second goal was to have resulting script look like it followed normal jQuery patterns, i.e. to see this get generated:

```
$("form").validate()
```

rather than

```
$.fn.validate($("form"))
```

Some examples:
Example of jQuery plugin that doesn't add to $.fn, but some APIs to $ directly:
https://github.com/nikhilk/scriptsharp/tree/master/src/Libraries/jQuery/jQuery.History

Example of jQuery plugins that do add to $.fn:
https://github.com/nikhilk/scriptsharp/tree/master/src/Libraries/jQuery/jQuery.Templating
https://github.com/nikhilk/scriptsharp/blob/master/samples/PhotoGrid/Plugins/LightBox.cs
https://github.com/nikhilk/scriptsharp/blob/master/samples/PhotoGrid/Plugins/Isotope.cs

Commit:
jQueryValidation now inherits from jQueryObject to support Plugin<T> pattern. Methods are non-static and have removed redundant parameters

Commit:
Renamed jQueryValidation to jQueryValidationObject to keep inline with other plugin naming conventions

Commit:
Renamed from jQueryValidation.cs

stevehobbsdev commented:
Ah. Unfortunately this has come about because I didn't actually realise the Plugin<T> pattern existed! It is indeed the better way as it is closer to the final state, as you say, and still doesn't require a cast.

No matter, it's a quick change. I've updated the code so that jQueryValidation inherits from jQueryObject. I've also renamed it so that it keeps inline with other plugin naming schemes.

Now we can achieve this:
`jQuery.Select("form").Plugin<jQueryValidationObject>().Validate(options);`

which produces this on the client:
`$("form").validate(options)`
